File: garden/seeds/java/app/src/test/java/com/WATER_ME_WITH_COMPANY/android/WATER_ME_WITH_APP_NAME/fragment/PresentableFragmentTest.java
package com.{{company_name}}.android.{{app_package_name_prefix}}.fragment;

import android.os.Bundle;

import com.{{company_name}}.android.{{app_package_name_prefix}}.R;
import com.{{company_name}}.android.{{app_package_name_prefix}}.{{app_class_prefix}}App;
import com.{{company_name}}.android.{{app_package_name_prefix}}.{{app_class_prefix}}TestRunner;
import com.{{company_name}}.android.{{app_package_name_prefix}}.module.AppServicesComponent;
import com.{{company_name}}.android.{{app_package_name_prefix}}.mvp.presenter.Presenter;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.shadows.support.v4.SupportFragmentTestUtil;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

@RunWith({{app_class_prefix}}TestRunner.class)
public class PresentableFragmentTest {

    @Test(expected = IllegalStateException.class)
    public void testThrowsExceptionIfNoPresenterProvided() {
        DummyFragment frag = DummyFragment.withNullPresenter();
        SupportFragmentTestUtil.startFragment(frag);
        frag.onCreate(mock(Bundle.class));
    }

    @Test
    public void testDelegatesToPresenter() {
        DummyFragment frag = DummyFragment.withMockPresenter();

        SupportFragmentTestUtil.startFragment(frag);

        verify(frag.presenter).onCreate(any(Bundle.class));
        verify(frag.presenter).onStart();
        verify(frag.presenter).onResume();

        frag.onSaveInstanceState(mock(Bundle.class));
        verify(frag.presenter).onSaveInstanceState(any(Bundle.class));

        frag.onPause();
        verify(frag.presenter).onPause();

        frag.onStop();
        verify(frag.presenter).onStop();

        frag.onDestroy();
        verify(frag.presenter).onDestroy();
    }


    public static class DummyFragment extends PresentableFragment<Presenter> {

        static DummyFragment withMockPresenter() {
            DummyFragment frag = withNullPresenter();
            frag.presenter = mock(Presenter.class);
            return frag;
        }

        static DummyFragment withNullPresenter() {
            DummyFragment frag = new DummyFragment();
            frag.app = ({{app_class_prefix}}App) RuntimeEnvironment.application;
            return frag;
        }

        Presenter presenter;

        @Override
        protected Presenter createPresenter(AppServicesComponent servicesComponent, Bundle savedInstanceState) {
            return presenter;
        }

        @Override
        protected int getLayoutId() {
            return R.layout.act_home;
        }
    }
}

File: garden/seeds/kotlin/app/src/test/kotlin/com/WATER_ME_WITH_COMPANY/android/WATER_ME_WITH_APP_NAME/fragment/PresentableFragmentTest.kt
package com.{{company_name}}.android.{{app_package_name_prefix}}.fragment


import android.os.Bundle
import com.{{company_name}}.android.{{app_package_name_prefix}}.{{app_class_prefix}}TestRunner
import com.{{company_name}}.android.{{app_package_name_prefix}}.TestUtils
import com.{{company_name}}.android.{{app_package_name_prefix}}.module.AppServicesComponent
import com.{{company_name}}.android.{{app_package_name_prefix}}.mvp.presenter.Presenter
import com.{{company_name}}.android.{{app_package_name_prefix}}.mvp.view.MvpView
import junit.framework.Test
import org.assertj.core.api.Assertions.assertThat
import org.junit.*
import org.junit.Test
import org.junit.Test
import org.junit.Test
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Matchers.any
import org.mockito.Mockito.mock
import org.mockito.Mockito.spy
import org.mockito.Mockito.verify
import org.robolectric.shadows.support.v4.SupportFragmentTestUtil

RunWith({{app_class_prefix}}TestRunner::class)
public class PresentableFragmentTest {

    Test
    public fun testDelegatesToPresenter() {
        val frag = DummyFragment()

        SupportFragmentTestUtil.startFragment(frag)

        val presenter = frag.presenter

        verify(presenter).onCreate(any(javaClass<Bundle>()))
        verify(presenter).onStart()
        verify(presenter).onResume()

        frag.onSaveInstanceState(mock(javaClass<Bundle>()))
        verify(presenter).onSaveInstanceState(any(javaClass<Bundle>()))

        frag.onPause()
        verify(presenter).onPause()

        frag.onStop()
        verify(presenter).onStop()

        frag.onDestroy()
        verify(presenter).onDestroy()
    }

    interface DummyMvpView : MvpView {

    }

    open class DummyPresenter(view: DummyMvpView, component : AppServicesComponent) : Presenter<DummyMvpView>(view, component) {

    }

    public class DummyFragment : PresentableFragment<DummyMvpView, DummyPresenter>() {

        override protected fun createPresenter(servicesComponent: AppServicesComponent, savedInstanceState: Bundle?): DummyPresenter {
            return spy(DummyPresenter(TestUtils.createView(javaClass<PresentableFragmentTest.DummyMvpView>()), servicesComponent))
        }

        override protected fun getLayoutId(): Int {
            return 0
        }
    }
}